src/main/scala/rifs/business/tables/ApplicationTables.scala

```
@@ -74,34 +75,43 @@ class ApplicationTables @Inject()(dbConfigProvider: DatabaseConfigProvider)(impl
-          case false => Future.successful(1)
-        }
-      case None =>
-        db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+    fetchAppWithSection(id, sectionNumber).flatMap {
+      case Some((app, Some(section))) => areDifferent(section.answers, answers) || completedAt.isDefined match {
+        case true => db.run(appSectionC(id, sectionNumber).update(section.copy(answers = answers, completedAt = completedAt)))
+        case false => Future.successful(1)
+      }
+      case Some((app, None)) => db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+      case None => Future.successful(0)
     }
   }
 
-  override def deleteSection(id: ApplicationId, sectionNumber: Int): Future[Int] = db.run {
-    appSectionC(id, sectionNumber).delete
+  def areDifferent(obj1: JsObject, obj2: JsObject): Boolean = {
+    val flat1 = JsonHelpers.flatten("", obj1).filter(_._2.trim != "")
+    val flat2 = JsonHelpers.flatten("", obj2).filter(_._2.trim != "")
+    flat1 != flat2
   }
 
-  def preCheckSaveAnswers(obj1: JsObject, obj2: JsObject, completedAt: Option[LocalDateTime]): Boolean = {
+  def joinedAppWithSection(id: Rep[ApplicationId], sectionNumber: Rep[Int]) = for {
+    as <- applicationTable joinLeft applicationSectionTable on ((a, s) => a.id === s.applicationId && s.sectionNumber === sectionNumber) if as._1.id === id
+  } yield as
 
-    !(JsonHelpers.flatten("", obj1).filter(_._2.isEmpty == false).toList.sortBy(_._2)
-      .equals(JsonHelpers.flatten("", obj2).filter(_._2.isEmpty == false).toList.sortBy(_._2)) &&
-      completedAt.isEmpty)
+  def appWithSectionQ(id: Rep[ApplicationId], sectionNumber: Rep[Int]) = joinedAppWithSection(id, sectionNumber)
+
+  lazy val appWithSectionC = Compiled(appWithSectionQ _)
+
+  override def deleteSection(id: ApplicationId, sectionNumber: Int): Future[Int] = db.run {
```

The calling code doesn't check if anything was actually deleted. Are we going to handle or ignore errors?